zebra-state/src/memory_state.rs

```diff
@@ -1,3 +1,406 @@
+use std::{
+    cmp::Ordering,
+    collections::BTreeSet,
+    collections::{BTreeMap, HashMap, HashSet},
+    ops::Deref,
+    sync::Arc,
+};
+
+use zebra_chain::{
+    block::{self, Block},
+    primitives::Groth16Proof,
+    sapling, sprout, transaction, transparent,
+    work::difficulty::PartialCumulativeWork,
+    work::difficulty::Work,
+};
+
+use crate::{service::QueuedBlock, BoxError};
+
 pub struct MemoryState {
     // TODO
 }
+
+struct ChainSet {
+    chains: BTreeSet<Chain>,
+
+    queued_blocks: BTreeMap<block::Hash, QueuedBlock>,
+    queued_by_parent: BTreeMap<block::Hash, Vec<block::Hash>>,
+    queued_by_height: BTreeMap<block::Height, Vec<block::Hash>>,
+}
```

Comment:
It seems like `MemoryState` and `ChainSet` are ultimately two names for the same type, wondering if we should remove the `MemoryState` type or rename the `ChainSet` type and update the RFC.

Reply:
What types will we need to implement contextual validation?

Here's one alternative:

/// A queue of blocks, awaiting the arrival of parent blocks.
pub struct QueuedBlocks {
    /// Blocks awaiting their parent blocks for contextual verification.
    blocks: BTreeMap<block::Hash, QueuedBlock>,
    /// Hashes from `queued_blocks`, indexed by parent hash.
    by_parent: BTreeMap<block::Hash, Vec<block::Hash>>,
    /// Hashes from `queued_blocks`, indexed by block height.
    by_height: BTreeMap<block::Height, Vec<block::Hash>>,
}

/// The state of the chains in memory, incuding queued blocks.
struct ChainState {
    /// Verified, non-finalized chains.
    chains: BTreeSet<Chain>,
    /// Blocks awaiting their parent blocks for contextual verification.
    contextual_queue: QueuedBlocks,
}

Here's another:

```suggestion
/// The state of the chains in memory, incuding queued blocks.
struct ChainState {
    /// Verified, non-finalized chains.
    chains: BTreeSet<Chain>,

    /// Blocks awaiting their parent blocks for contextual verification.
    queued_blocks: BTreeMap<block::Hash, QueuedBlock>,
    /// Hashes from `queued_blocks`, indexed by parent hash.
    queued_by_parent: BTreeMap<block::Hash, Vec<block::Hash>>,
    /// Hashes from `queued_blocks`, indexed by block height.
    queued_by_height: BTreeMap<block::Height, Vec<block::Hash>>,
}
```